core/src/main/java/org/bitcoinj/core/AbstractBlockChain.java

```diff
@@ -672,7 +672,8 @@ private void handleNewBestChain(StoredBlock storedPrev, StoredBlock newChainHead
                     txOutChanges = connectTransactions(cursor);
                 else
                     txOutChanges = connectTransactions(newChainHead.getHeight(), block);
-                storedNewHead = addToBlockStore(storedNewHead, cursor.getHeader(), txOutChanges);
+                storedNewHead = addToBlockStore(storedNewHead,
+                        cursor.getHeader().transactions == null ? cursor.getHeader() : cursor.getHeader().cloneAsHeader(), txOutChanges);
```

Comment:
Is the branch ever taken here? We call connectTransactions just above it, so It hink "cursor.getHeader().transactions == null" is always false. This should just be cloneAsHeader()

You could also pull out the repeated cursor.getHeader() calls into a variable here. IntelliJ can do it for you automatically.